### ISSUE TITLE
Different file filter for different file types

### ISSUE THREAD

**Issue by limzykenneth** [expressjs/multer#474]
I'm trying to create a form with two upload fields, one accepting only images and the other accepts any type of file. However, after trying out lots of potential solution I can think of, I can't find a solution to validate the two fields separately.

Essentially I need to have two different multer settings, one for the image upload (validating the mime type and saving in `uploads/images` folder) the other for file upload (no validating mime type and saving in `uploads/files` folder).

```javascript
function createMulterMiddleware(fields, destination, mimeType = [], maxSize = 20000000){
	// fields = [{name: "name", maxCount: count}]
	console.log(fields);
	var setting = multer({
		storage: multer.diskStorage({
			destination: function (req, file, cb) {
				cb(null, path.join("./uploads", destination));
			},
			filename: function (req, file, cb) {
				cb(null, path.basename(file.originalname, path.extname(file.originalname)) + '-' + randomstring.generate(10) + path.extname(file.originalname));
			}
		}),
		fileFilter: function(req, file, cb){
			if(_.includes(mimeType, file.mimetype) || mimeType.length === 0){
				cb(null, true);
			}else{
				cb(null, false);
			}
		},
		limits: {
			fileSize: maxSize
		}
	}).fields(fields);

	return setting;
}

uploadSchemas.middleware = function(req, res, next){
	var requestFields = _.find(res.locals.schemas, {collectionSlug: req.params.collection}).fields;

	_.each(requestFields, function(el, i){
		if(el.type == "image") {
			console.log("image");
			createMulterMiddleware([{
				name: el.slug,
				maxCount: 1
			}], "images", ["image/gif", "image/jpeg", "image/png"])(req, res, function(err){
				if(err) throw err;
			});

		}else if(el.type == "file"){
			console.log("file");
			createMulterMiddleware([{
				name: el.slug,
				maxCount: 1
			}], "files")(req, res, function(err){
				if(err) throw err;
			});
		}
	});
	next();
};
```

The above will fail because the multer middleware will look at all the `file` input field and throws because it found a field not accounted for for that particular middleware, but is actually handled by the next one.

In short, is there anyway to have separate multer settings for the same route? I'm not sure if I'm being clear about what my problem is so feel free to ask me for more info, really appreciate any help here as I've tried to solve this for the past few days with no luck. Thanks!

**Comment by ComputerCarl** [comment]
I know this is an old question, but I hope this helps;
Why have two form fields for files? Relyimg pn an end user to put stuff in different fields surely can't end well.

In the configuration on `multer`, you can configure the storage location based on criteria;
````js
const uuid = require('uuid');
const multer = require('multer');

// configure multer service with custom filename
// https://github.com/expressjs/multer
const multerStorage = multer.diskStorage({
  destination: (req, file, cb) =>{
    console.log(file);
    // set destination based on tests of file, e.g. file.mimetype
    var dest = (0 > 1) ? 'public/media' : 'public/images'
    cb(null, dest)
  },
  filename: (req, file, cb) =>{
    cb(null, uuid() + file.originalname.match(/\.\w+$/));
  }
});
const multerUpload = multer({storage: multerStorage});
````
